### PR TITLE
add posefactor expression

### PR DESCRIPTION
**Pull request by yetongumich**
- Add poseConstraint expression
- Replace PoseFactor with an ExpressionFactor

### CODE REVIEW

File: gtdynamics/universal_robot/Joint.cpp [hunk 1]
```diff
@@ -295,4 +297,25 @@ std::ostream &operator<<(std::ostream &os, const JointSharedPtr &j) {
...
+  Pose3_ wTp(internal::PoseKey(parent()->id(), t));
+  Pose3_ wTc(internal::PoseKey(child()->id(), t));
+  gtsam::Double_ q(internal::JointAngleKey(id(), t));
+
+  // Compute the expected pose of the child link.
+  Pose3_ pTc(std::bind(&Joint::parentTchild, this, _1, _2), q);
+  Pose3_ wTc_hat = compose(wTp, pTc);
+
+  // Return the error in tangent space
+  return logmap(wTc, wTc_hat);
```
Comment (dellaert) [inline gtdynamics/universal_robot/Joint.cpp:318]: just add a lomap2 for now, with a TODO in code and issue to remove it.

File: gtdynamics/factors/PoseFactor.h [hunk 2]
```diff
@@ -37,10 +38,10 @@ using boost::assign::cref_list_of;
  * Given the joint model, this factor optimizes for the underlying joint axis
  * and the corresponding poses of the parent and child links.
  */
-class PoseFactor : public gtsam::NoiseModelFactor {
+class PoseFactor : public gtsam::ExpressionFactor<gtsam::Vector6> {
```
Comment (dellaert) [inline gtdynamics/factors/PoseFactor.h:41]: replace with inline function `PoseFactor`
Add a note that this is provided for BW compatibility only, and will in time be replaced with EqualityConstraint

File: gtdynamics/universal_robot/Joint.h [hunk 3]
```diff
@@ -365,6 +366,13 @@ class Joint : public boost::enable_shared_from_this<Joint> {
     return isChildLink(link) ? parentPoseTwist(wTl, Vl, q, q_dot)
                              : childPoseTwist(wTl, Vl, q, q_dot);
   }
+
+  /**
+   * @brief Create expression that constraint the pose of two links imposed by
+   * the joint angle.
+   */
+  gtsam::Expression<gtsam::Vector6> poseConstraint(uint64_t t) const;
```
Comment (dellaert) [inline gtdynamics/universal_robot/Joint.h:374]: Add default 0

File: gtdynamics/universal_robot/Joint.cpp [hunk 4]
```diff
@@ -295,4 +297,25 @@ std::ostream &operator<<(std::ostream &os, const JointSharedPtr &j) {
   return j->to_stream(os);
 }
 
+/* ************************************************************************* */
+gtsam::Expression<gtsam::Vector6> Joint::poseConstraint(
+    uint64_t t) const {
+  using gtsam::Expression;
```
Comment (dellaert) [inline gtdynamics/universal_robot/Joint.cpp:303]: kill first 2 using

**Review by dellaert — APPROVED** [review]
Add logmap2 (inside .cpp) so CI will pass and merge at will.

File: gtdynamics/universal_robot/Joint.h (hunk 3 above)
Comment (yetongumich) [reply gtdynamics/universal_robot/Joint.h:374]: Done

File: gtdynamics/universal_robot/Joint.cpp (hunk 4 above)
Comment (yetongumich) [reply gtdynamics/universal_robot/Joint.cpp:303]: Done.

File: gtdynamics/universal_robot/Joint.cpp [hunk 5]
```diff
@@ -295,4 +297,25 @@ std::ostream &operator<<(std::ostream &os, const JointSharedPtr &j) {
...
+  using namespace std::placeholders;
+
+  // Get an expression for parent pose.
+  Pose3_ wTp(internal::PoseKey(parent()->id(), t));
+  Pose3_ wTc(internal::PoseKey(child()->id(), t));
+  gtsam::Double_ q(internal::JointAngleKey(id(), t));
+
+  // Compute the expected pose of the child link.
+  Pose3_ pTc(std::bind(&Joint::parentTchild, this, _1, _2), q);
+  Pose3_ wTc_hat = compose(wTp, pTc);
```
[the remark this answers is not in the document]
Comment (yetongumich) [reply gtdynamics/universal_robot/Joint.cpp:315]: Done

File: gtdynamics/universal_robot/Joint.cpp (hunk 1 above)
Comment (yetongumich) [reply gtdynamics/universal_robot/Joint.cpp:318]: Done

File: gtdynamics/factors/PoseFactor.h (hunk 2 above)
Comment (yetongumich) [reply gtdynamics/factors/PoseFactor.h:41]: Done

**Comment by dellaert** [comment]
Not all comments have been addressed yet, correct? After that, please feel free to merge

**Comment by varunagrawal** [comment]
Okay, I don't understand the rationale behind this PR. This is something that was not discussed with the rest of the group and as such there are ramifications to this change aka these are breaking changes. If we start doing this without proper discussion, then we're just going to end up with mass chaos (especially with deadlines looming).

1. The pose factor class has been removed, so my state estimation code no longer works.
2. @yetongumich simply commented out the wrapped portions of the PoseFactor and I don't think he tested it, because ExpressionFactors are not wrapped.
3. If there was a need for Expressions in the pose factor, there should be a factory function that converts the regular NonlinearFactor to the ExpressionFactor.

Honestly, I have to say that this PR was rushed and not well thought out with respect to the rest of the group.

**Pull request merged**